### PR TITLE
move annotations on header to fields

### PR DESCRIPTION
**Pull request by hanw**
This change will keep the annot on struct with the fields that the struct is flattened to.
```
@my_annot
struct st {
    bit<8> f0; 
}

header hd{
   st s;
}
```

```
header hd {
   @my_annot _s_f0;
}
```

### CODE REVIEW

**Comment by mbudiu-vmw** [comment]
You can't do this for all annotations; if you move for example `@name` annotations this will break the back-end which uses the original annotations or names to output the names.

**Comment by hanw** [comment]
I added code to avoid copying name annotation.

File: midend/flattenHeaders.cpp [hunk 1]
```diff
@@ -19,20 +19,31 @@ limitations under the License.
 
 namespace P4 {
 
+template <typename Func>
```
Comment (mbudiu-vmw) [inline midend/flattenHeaders.cpp]: The Annotations class already has a function called 'where' which does exactly this.

File: midend/flattenHeaders.h [hunk 2]
```diff
@@ -177,6 +178,7 @@ class FlattenHeaders final : public PassManager {
         passes.push_back(new FindHeaderTypesToReplace(sm));
         passes.push_back(new ReplaceHeaders(sm));
         passes.push_back(new ClearTypeMap(typeMap));
+        passes.push_back(new TypeChecking(refMap, typeMap, true));
```
Comment (mbudiu-vmw) [inline midend/flattenHeaders.h]: In general you should not end with type checking; if the next pass needs it, it should call it itself.

File: midend/flattenHeaders.cpp (hunk 1 above)
Comment (hanw) [reply midend/flattenHeaders.cpp]: fixed.

File: midend/flattenHeaders.h (hunk 2 above)
Comment (hanw) [reply midend/flattenHeaders.h]: fixed.

File: midend/flattenHeaders.cpp (hunk 1 above)
Comment (cc10512) [reply midend/flattenHeaders.cpp]: please remove this function then.

**Comment by hanw** [comment]
> This code will probably work, but it may need to be revisited if we add other annotations.
> It may also interfere with other back-end specific annotations.

I added a policy to this pass to allow backend to customize the annotation that it is willing to keep during flattening. By default, all annotation will be lost. Backend must implement the policy if it needs certain annotation to be kept from struct to fields.

**Pull request merged**